decisions/DR007-Hierarchical-or-compositional-traits-for-specifications.md

```diff
@@ -0,0 +1,600 @@
+    int endFrame
+    int size
+  }
+  class Panoramic {
+    float angle
+  }
+  class Shot {
+    int startFrame
+    int endFrame
+  }
+  class Container {
+    bool isEmpty
+  }
+```
+
+```
+  specification Blob {
+    str mimeType
+  }
+  specification Blob.Geometry {
+    str upAxis
+  }
+  specification Blob.Geometry.AnimationCache {
+    int startFrame
+    int endFrame
+  }
+  specification Blob.Geometry.Rig {
+    str department
+  }
+  specification Blob.Image {
+    int channels
+  }
+  specification Blob.Image.RenderCache {
+    int startFrame
+    int endFrame
+  }
+  specification Blob.Image.Panoramic {
+    float angle
+  }
+  specification Container {
+    bool isEmpty
+  }
+  specification Container.Shot {
+    int startFrame
+    int endFrame
+  }
+```
+
+If we change the hierarchy to share Sequence, Rig and Panoramic now have
+an unwanted startFrame and endFrame property that might not apply. In
+addition, all sequences have to be containers, or all containers have to
+be sequences.
+
+```mermaid
+classDiagram
+  direction TB
+  Blob <-- Geometry
+  Sequence <-- Blob
+  Geometry <-- AnimationCache
+  Geometry <-- Rig
+  Blob <-- Image
+  Image <-- RenderCache
+  Image <-- Panoramic
+  Sequence <-- Shot
+  Container <-- Sequence : everything has to be a container
+  
+  class Blob {
+    str mimeType
+  }
+  class Sequence {
+    int startFrame
+    int endFrame
+  }
+  class Geometry {
+    str upAxis
+  }
+  class AnimationCache {
+    int size
+  }
+  class Rig {
+    str department
+  }
+  class Image {
+    int channels
+  }
+  class RenderCache {
+    int size
+  }
+  class Panoramic {
+    float angle
+  }
+  class Shot {
+  }
+  class Container {
+    bool isEmpty
+  }
+```
+
+```mermaid
+classDiagram
+  direction TB
+  Blob <-- Geometry
+  Sequence <-- Blob
+  Geometry <-- AnimationCache
+  Geometry <-- Rig
+  Blob <-- Image
+  Image <-- RenderCache
+  Image <-- Panoramic
+  Container <-- Shot
+  Sequence <-- Container : all containers have to be sequences
+  
+  class Blob {
+    str mimeType
+  }
+  class Sequence {
+    int startFrame
+    int endFrame
+  }
+  class Geometry {
+    str upAxis
+  }
+  class AnimationCache {
+    int size
+  }
+  class Rig {
+    str department
+  }
+  class Image {
+    int channels
+  }
+  class RenderCache {
+    int size
+  }
+  class Panoramic {
+    float angle
+  }
+  class Shot {
+
+  }
+  class Container {
+    bool isEmpty
+  }
+```
+
+We could copy the sequence specification but since only one parent is
+possible, it has to be copied. If we don't copy the Sequence class, then
+it becomes multiple inheritance.
+
+```mermaid
+classDiagram
+  direction TB
+  Blob <-- Geometry
+  Geometry <-- Sequence_3
+  Geometry <-- Rig
+  Blob <-- Image
+  Sequence_1 <-- RenderCache
+  Image <-- Sequence_1
+  Image <-- Panoramic
+  Sequence_2 <-- Shot
+  Container <-- Sequence_2
+  Sequence_3 <-- AnimationCache
+  
+  class Blob {
+    str mimeType
+  }
+  class Sequence_1 {
```

Comment:
[consdier][Strong] In Option 1, they don't have to have unique names. So it can still be `Sequence` in all branches.

Reply:
Even though they can have the same names, they are different specifications because the hierarchy is involved. This could lead to confusion as `blob.geometry.sequence` is not the same as `blob.image.sequence` and not the same as `container.sequence` .  Calling it different names highlights that they are different copies of the same specification. Calling them all Sequence without the number makes it falsely seem like these are shared attributes from the same class. Mermaid will show this as causing a contradiction of each class having a single parent.

```mermaid
classDiagram
  direction TB
  Blob <-- Geometry
  Geometry <-- Sequence
  Geometry <-- Rig
  Blob <-- Image
  Sequence <-- RenderCache
  Image <-- Sequence
  Image <-- Panoramic
  Sequence <-- Shot
  Container <-- Sequence
  Sequence <-- AnimationCache
  
  class Blob {
    str mimeType
  }
  class Sequence {
    int startFrame
    int endFrame
  }

  
  class Geometry {
    str upAxis
  }
  class AnimationCache {
    int size
  }
  class Rig {
    str department
  }
  class Image {
    int channels
  }
  class RenderCache {
    int size
  }
  class Panoramic {
    float angle
  }
  class Shot {

  }
  class Container {
    bool isEmpty
  }
```

Reply:
I think there might be some crossed wires. In the Option 1, the "Sequence" class as its depicted in your diagram does not exist at all, and there are no copies of the specification. The  _only_ things that exist are concrete Specifications with complete schema strings such as `blob.geometry.sequence`,  `container.sequence` etc. There is no manifestation of the individual partial intermediate schema name components, _only_ parent schemas that have the same exact prefix as some part of the name. So there is no need to disambiguate - the resulting `Specification` is already uniquely distinguished by the full schema name. So when you say "calling it sequence", "it" is referring to something that doesn't actually exist in the model. 

In the class hierarchy, there is only `BlobSpecification("blob")`, `GeometrySpecification("blob.geometry")`, `GeometrySequenceSpecificaiton("blob.geomety.sequence")`, `ImageSequenceSpecification("blob.image.sequence")`. Where `GeometrySpecification` is derived from `BlobSpecification`, etc. Consequently, there is never any ambiguity as the hierarchical namespace ensures this.

This is the main reason the trait approach is a) different and b) better suited for purpose - the explicit lack of any concrete, shared `sequence` concept _in_ the hierarchical model is why it is bad, and why we want option 2. In Option 2 it only exists as an emergent phenomena though commonly named sub-strings. If you do end up uniquing the component name, it actually ends up evaporating that already tenuous link further still.

I think this might be a side effect of using a node graph diagram to represent the hierarchical model in this way. It depicts (and consequently implies) the existence of intermediate specifications as distinct entities within the model. In reality, it is only _visualisation_ of the implicit hierarchy in the resulting scheme strings themselves, not a reflection of the actual model structure. The intermediate entities in the path don't actually exist in Option 1 at all. This is what results in the diagram noted above then being misleading - to represent the actual model conceptually, it would need three `Sequence` blocks all with the same name - the schema component names would never be suffixed in the specification itself. Ie.the schema name for `GeometrySequence` it would never be `blob.geometry.sequence_2` - that implies some rule that name components are globally unique, defeating the purpose of the namespace hierarchy in the first place.

Reply:
Using nested blocks for Option 1 would probably be a more accurate representation, where an outer block represent a parent name component?

Reply:
The point of the diagram is to show that `Sequence` can't be shared between branches of the hierarchy and be the same specification, not that it's the best case representation of Option 1. The first diagram is the representation of Option 1, this diagram is to show that it's not possible to share the same specification without joining the hierarchies together. I could make a diagram that illustrates the naming rules, but that wouldn't illustrate the problem of having to join the branches together as well.